### PR TITLE
serial interface: add command parser for protocol

### LINKED ISSUE #60: serial interface: add command parser for protocol
Create a command parser which is capable to interpret the commands from serial interface according to the [defined protocol](https://github.com/Task-Tracker-Systems/Protocol-Serial-Interface/blob/631fc844a80ad5d707d14fb060660ac1e2c31cdd/serial-protocol.md#protocol-v001).

See [WIP](https://github.com/Task-Tracker-Systems/Task-Tracker-Device/tree/aa8496b3a6a37e3139d315d498f1a5c19dcb29ec/lib/serial_interface).

Acceptance criteria:

- It is OK for this version to respond with dummy responses (format does not need to follow protocol)
- All arguments of the commands shall be understood (test by echo the values to serial interface)
- Dummy functions for the commands shall be called (test by echo a string on serial interface)

Depends on: 

- #58
- #59 or #77

### PR DESCRIPTION
**Pull request by dhebbeker**
### Purpose

Create a command parser which is capable to interpret the commands from serial interface according to the [defined protocol](https://github.com/Task-Tracker-Systems/Protocol-Serial-Interface/blob/631fc844a80ad5d707d14fb060660ac1e2c31cdd/serial-protocol.md#protocol-v001).

Resolves #60 

### Acceptance criteria

- [x] It is OK for this version to respond with dummy responses (format does not need to follow protocol)
- [x] All arguments of the commands shall be understood (test by echo the values to serial interface)
- [x] Dummy functions for the commands shall be called (test by echo a string on serial interface)

### Todos

- [x] remove trim functionality as it is not used
- [x] add member function to `Command` which generates a string which is intended to specify the correct use of the command (command line format)
- [x] handle the case that the command line can not be correctly interpreted outside of `Command` by printing a message 
- [x] refactoring: check which function shall be member functions
- [x] add documentation
- [x] test: add appropriate test cases (check for code coverage)
- [x] create dedicated namespace
- [x] add test case for string helper functions
- [x] add documentation for string helper functions
- [x] add documentation on how to use the command line interpreter

### Changes

- `.github/workflows/platformio.yml`: Added test execution to workflow.
- `lib/3rd_party_adapters/serial_port.cpp`: Change line extraction to omit line ending characters and empty lines.
- `lib/application_business_rules/Protocol.*`: This implements the protocol (here: dummies) by using the command line interpreter.
- `lib/application_business_rules/command_line_interpreter.hpp`: Essentially provides `Option` and `Command` class templates which can be used to call a function with some arguments based on a command line.
- `lib/enterprise_business_rules/string_helpers.*`: Add helper functions which can be used to operate on strings.
- `platformio.ini`: Add configuration for unit tests.
- `src/main.cpp`: Attach `ProtocolHandler` to serial port.
- `test/test_cli/test_cli.cpp`: Unit test for command line interpreter.
- `test/test_serial_port/test_serial_port.cpp`: Adjust to changed fetching of lines. The end of line character is omitted now.
- `test/test_string_helpers/test_string_helpers.cpp`: Unit test for string helpers.

### CODE REVIEW

**Review by dhebbeker — COMMENTED** [review]
Did a self-review: Some documentation is missing.

**Review by dhebbeker — COMMENTED** [review]
Did self-review: OK

**Comment by dhebbeker** [comment]
Tested also with real device:

```plain
 begin program '/home/dhebbeker/pro/time-tracker/src/time-tracker.ino compiled at Nov 30 2023 21:50:08
---- Sent utf8 encoded message: "list\n" ----
this is a list: a, b, c, ...
---- Sent utf8 encoded message: "list\n" ----
this is a list: a, b, c, ...
---- Sent utf8 encoded message: "edit --name \"hello world\"\n" ----
Edit id(0) label('hello world') duration(0)
---- Sent utf8 encoded message: "edit --name \"@²³ß\\\"ä\" --duration 1337\n" ----
Edit id(0) label('@²³ß"ä') duration(1337)
```
:heavy_check_mark:

**Review by FiveTeethless — DISMISSED** [review]
looks ok for me.
I didn't go through all the changes, but it looks promising ;)

File: test/test_serial_port/test_serial_port.cpp [hunk 1]
```diff
@@ -62,7 +62,7 @@ void test_subscribeToLine()
         When(Method(ArduinoFake(Serial), read)).Return(testLine[i]);
         serialEvent();
     }
-    TEST_ASSERT_EQUAL_STRING(testLine, receivedLine.c_str());
+    TEST_ASSERT_EQUAL_STRING("C++", receivedLine.c_str());
```
Comment (FiveTeethless) [inline test/test_serial_port/test_serial_port.cpp:65]: was this change intentional?
I just found this, as I tried to find a solution for the conflict in the platformio.yml.

Comment (dhebbeker) [reply test/test_serial_port/test_serial_port.cpp:65]: Yes this change was intentional: The algorithm for reading lines has changed as such that the end-of-line characters are now discarded. That is why `testLine = "C++\n"` can not be used for comparison any more.

**Review by dhebbeker — COMMENTED** [review]
I did rebase the branch in order to resolve merge conflicts. That means that an updated review is required.
